File: products/opensyber/packages/tokenforge-sdks/kotlin/TokenForge.kt
/**
 * TokenForge Kotlin SDK — Device-bound ECDSA P-256 session security for Android.
 */
package cloud.opensyber.tokenforge

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.security.*
import java.security.spec.ECGenParameterSpec
import java.util.Base64
import java.util.UUID

class TokenForge(
    private val apiKey: String,
    private val apiBase: String = "https://tokenforge-api.opensyber.cloud",
) {
    val deviceId: String = UUID.randomUUID().toString()
    val sessionId: String = UUID.randomUUID().toString()
    private val keyPair: KeyPair = loadOrGenerateKey()
    private val client = OkHttpClient()

    private fun loadOrGenerateKey(): KeyPair {
        val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        if (ks.containsAlias(KEYSTORE_ALIAS)) {
            val entry = ks.getEntry(KEYSTORE_ALIAS, null) as KeyStore.PrivateKeyEntry
            return KeyPair(entry.certificate.publicKey, entry.privateKey)
        }
        val spec = KeyGenParameterSpec.Builder(
            KEYSTORE_ALIAS,
            KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY,
        )
            .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
            .setDigests(KeyProperties.DIGEST_SHA256)
            .setUserAuthenticationRequired(false)
            .build()
        val generator = KeyPairGenerator.getInstance(
            KeyProperties.KEY_ALGORITHM_EC, "AndroidKeyStore",
        )
        generator.initialize(spec)
        return generator.generateKeyPair()
    }

    private fun sign(payload: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(payload.toByteArray())
        val signer = Signature.getInstance("NONEwithECDSA").apply {
            initSign(keyPair.private)
            update(digest)
        }
        return Base64.getEncoder().encodeToString(signer.sign())
    }

    fun getHeaders(): Map<String, String> {
        val nonce = UUID.randomUUID().toString().replace("-", "")
        val timestamp = (System.currentTimeMillis() / 1000).toString()
        val payload = "$sessionId:$nonce:$timestamp"
        return mapOf(
            "X-TF-Signature" to sign(payload),
            "X-TF-Nonce" to nonce,
            "X-TF-Timestamp" to timestamp,
            "X-TF-Device-ID" to deviceId,
            "Authorization" to "Bearer $apiKey",
        )
    }

    fun signRequest(request: Request): Request {
        val builder = request.newBuilder()
        getHeaders().forEach { (k, v) -> builder.header(k, v) }
        return builder.build()
    }

    suspend fun bind(): String = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("deviceId", deviceId)
            put("sessionId", sessionId)
            put("publicKey", publicKeyPem())
        }
        val request = Request.Builder()
            .url("$apiBase/v1/bind")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()
        val signed = signRequest(request)
        val response = client.newCall(signed).execute()
        check(response.isSuccessful) { "Bind failed: ${response.code}" }
        response.body?.string() ?: ""
    }

    fun interceptor(): Interceptor = Interceptor { chain ->
        val signed = signRequest(chain.request())
        chain.proceed(signed)
    }

    fun okHttpClient(): OkHttpClient {
        return OkHttpClient.Builder().addInterceptor(interceptor()).build()
    }

    private fun publicKeyPem(): String {
        val encoded = Base64.getMimeEncoder(64, "\n".toByteArray())
            .encodeToString(keyPair.public.encoded)
        return "-----BEGIN PUBLIC KEY-----\n$encoded\n-----END PUBLIC KEY-----"
    }

    companion object {
        private const val KEYSTORE_ALIAS = "tokenforge_device_key"
    }
}

File: products/opensyber/packages/tokenforge-sdks/swift/TokenForge.swift
/// TokenForge Swift SDK — Device-bound ECDSA P-256 session security for iOS.
/// Prefers Secure Enclave (hardware-backed) keys with automatic software fallback.
import Foundation
import Security
import CryptoKit

/// Wraps both Secure Enclave and software-backed P-256 signing keys.
enum TokenForgeKey {
    case software(P256.Signing.PrivateKey)
    case secureEnclave(SecureEnclave.P256.Signing.PrivateKey)

    var isHardwareBacked: Bool {
        if case .secureEnclave = self { return true }
        return false
    }

    func sign(_ data: Data) throws -> P256.Signing.ECDSASignature {
        switch self {
        case .software(let key): return try key.signature(for: data)
        case .secureEnclave(let key): return try key.signature(for: data)
        }
    }

    var publicKey: P256.Signing.PublicKey {
        switch self {
        case .software(let key): return key.publicKey
        case .secureEnclave(let key): return key.publicKey
        }
    }
}

public final class TokenForge: @unchecked Sendable {
    private let apiKey: String
    private let apiBase: String
    private let sessionId: String
    private let deviceId: String
    private let key: TokenForgeKey
    private static let keychainTag = "cloud.opensyber.tokenforge.key".data(using: .utf8)!

    /// Whether the signing key is backed by the Secure Enclave hardware.
    public var isHardwareBacked: Bool { key.isHardwareBacked }

    public init(apiKey: String, apiBase: String = "https://tokenforge-api.opensyber.cloud") {
        self.apiKey = apiKey
        self.apiBase = apiBase
        self.sessionId = UUID().uuidString
        self.deviceId = UUID().uuidString
        self.key = Self.loadOrGenerateKey()
    }

    // MARK: - Key Generation & Storage

    private static func loadOrGenerateKey() -> TokenForgeKey {
        if let existing = loadExistingKey() { return existing }
        if SecureEnclave.isAvailable, let seKey = try? SecureEnclave.P256.Signing.PrivateKey() {
            saveSecureEnclaveFlag(true)
            return .secureEnclave(seKey)
        }
        let key = P256.Signing.PrivateKey()
        saveSoftwareKey(key)
        return .software(key)
    }

    private static func loadExistingKey() -> TokenForgeKey? {
        if loadSecureEnclaveFlag(),
           let seKey = try? SecureEnclave.P256.Signing.PrivateKey() {
            return .secureEnclave(seKey)
        }
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keychainTag,
            kSecReturnData as String: true,
        ]
        var result: AnyObject?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
              let data = result as? Data,
              let key = try? P256.Signing.PrivateKey(rawRepresentation: data) else { return nil }
        return .software(key)
    }

    private static func saveSoftwareKey(_ key: P256.Signing.PrivateKey) {
        let attrs: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: keychainTag,
            kSecValueData as String: key.rawRepresentation,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
        ]
        SecItemAdd(attrs as CFDictionary, nil)
    }

    /// SE keys cannot be exported -- persist a flag so we know to reload from hardware.
    private static func saveSecureEnclaveFlag(_ exists: Bool) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "cloud.opensyber.tokenforge",
            kSecAttrAccount as String: "se-key-flag",
        ]
        SecItemDelete(query as CFDictionary)
        var attrs = query
        attrs[kSecValueData as String] = Data(exists ? [1] : [0])
        attrs[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        SecItemAdd(attrs as CFDictionary, nil)
    }

    private static func loadSecureEnclaveFlag() -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "cloud.opensyber.tokenforge",
            kSecAttrAccount as String: "se-key-flag",
            kSecReturnData as String: true,
        ]
        var result: AnyObject?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
              let data = result as? Data, data.first == 1 else { return false }
        return true
    }

    // MARK: - Signing

    private func sign(payload: String) -> String {
        let data = Data(SHA256.hash(data: Data(payload.utf8)))
        guard let signature = try? key.sign(data) else { return "" }
        return Data(signature.derRepresentation).base64EncodedString()
    }

    public func signedHeaders() -> [String: String] {
        let nonce = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        let timestamp = String(Int(Date().timeIntervalSince1970))
        let payload = "\(sessionId):\(nonce):\(timestamp)"
        return [
            "X-TF-Signature": sign(payload: payload),
            "X-TF-Nonce": nonce,
            "X-TF-Timestamp": timestamp,
            "X-TF-Device-ID": deviceId,
            "Authorization": "Bearer \(apiKey)",
        ]
    }

    public func signRequest(_ request: URLRequest) -> URLRequest {
        var req = request
        for (header, value) in signedHeaders() { req.setValue(value, forHTTPHeaderField: header) }
        return req
    }

    // MARK: - Device Binding

    public func bind() async throws -> Data {
        let url = URL(string: "\(apiBase)/v1/bind")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "deviceId": deviceId, "sessionId": sessionId,
            "publicKey": publicKeyPEM(), "hardwareBacked": isHardwareBacked,
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let signed = signRequest(request)
        let (data, response) = try await URLSession.shared.data(for: signed)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw TokenForgeError.bindFailed
        }
        return data
    }

    private func publicKeyPEM() -> String {
        let base64 = key.publicKey.derRepresentation.base64EncodedString(options: .lineLength64Characters)
        return "-----BEGIN PUBLIC KEY-----\n\(base64)\n-----END PUBLIC KEY-----"
    }

    public func registerInterceptor() {
        TokenForgeURLProtocol.tokenForge = self
        URLProtocol.registerClass(TokenForgeURLProtocol.self)
    }
}

public enum TokenForgeError: Error { case bindFailed }

public final class TokenForgeURLProtocol: URLProtocol {
    static var tokenForge: TokenForge?
    override public class func canInit(with request: URLRequest) -> Bool {
        URLProtocol.property(forKey: "TFSigned", in: request) == nil
    }

    override public class func canonicalRequest(for request: URLRequest) -> URLRequest { request }
    override public func startLoading() {
        guard let tf = Self.tokenForge else { return }
        var signed = tf.signRequest(request)
        URLProtocol.setProperty(true, forKey: "TFSigned", in: signed as! NSMutableURLRequest)
        let task = URLSession.shared.dataTask(with: signed) { data, response, error in
            if let error { self.client?.urlProtocol(self, didFailWithError: error); return }
            if let response { self.client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed) }
            if let data { self.client?.urlProtocol(self, didLoad: data) }
            self.client?.urlProtocolDidFinishLoading(self)
        }
        task.resume()
    }

    override public func stopLoading() {}
}
